ambry-router/src/main/java/com/github/ambry/router/ParanoidDurabilityOperationTracker.java

```diff
@@ -0,0 +1,405 @@
+        // originating DC, we can be sure the operation will end up with a NOT_FOUND error.
+        if (trackedRequestFinalState == TrackedRequestFinalState.NOT_FOUND) {
+          totalNotFoundCount++;
+          if (replicaId.getDataNodeId().getDatacenterName().equals(originatingDcName)) {
+            if (originatingDcLeaderOrStandbyReplicas.contains(replicaId)) {
+              // Since bootstrap replicas could still be catching up, treat NotFound responses only from leader or
+              // standby replicas as valid.
+              originatingDcNotFoundCount++;
+            }
+          }
+        }
+    }
+  }
+
+  public boolean hasFailed() {
+    if (routerConfig.routerPutUseDynamicSuccessTarget) {
+      for(String dcName : allDatacenters) {
+        if (dcName.equals(datacenterName)) {  // Local colo
+          if (hasFailedDynamically(dcName, routerConfig.routerPutSuccessTarget)) return true;
+        } else {
+          if (hasFailedDynamically(dcName, routerConfig.routerPutRemoteSuccessTarget)) return true;
+        }
+      }
+      return false;
+    } else {
+      return hasFailedLocally() || hasFailedRemotely();
+    }
+  }
+
+  private boolean hasFailedDynamically(String dc, int successTarget)  {
+    return (totalReplicaCountByDc.get(dc) - failedCountByDc.get(dc)) < Math.max(totalReplicaCountByDc.get(dc) - 1, successTarget + disabledCountByDc.get(dc));
+  }
+
+  private boolean hasFailedRemotely() {
+    for(String dcName : allDatacenters) {
+      if(!dcName.equals(datacenterName)) {
+        if (replicaSuccessCountByDc.get(dcName) < Math.max(totalReplicaCountByDc.get(dcName) - 1,
+            routerConfig.routerPutRemoteSuccessTarget + disabledCountByDc.get(dcName))) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  private boolean hasFailedLocally() {
+    return replicaInPoolOrFlightCountByDc.get(datacenterName) + replicaSuccessCountByDc.get(datacenterName) < replicaSuccessTarget;
+  }
+
+  /**
+   * Add {@code delta} to a replicas in pool or in flight counter.
+   * @param delta the value to add to the counter.
+   */
+  private void modifyReplicasInPoolOrInFlightCount(ReplicaId replica, int delta) {
+    String dcName = replica.getDataNodeId().getDatacenterName();
+    replicaInPoolOrFlightCountByDc.put(dcName, replicaInPoolOrFlightCountByDc.get(dcName) + delta);
+  }
+
+  // Is it OK to use this method instead of the getCurrentParallelism() method in SimpleOperationTracker?
+  // The name is different but not the contents.
+  int getCurrentLocalParallelism() {
+    return replicaParallelism;
+  }
+
+  int getCurrentRemoteParallelism() {
+    return remoteReplicaParallelism;
+  }
+
+  @Override
+  public Iterator<ReplicaId> getReplicaIterator() {
+    replicaIterator = replicaPool.iterator();                  //Iterator<ReplicaId>, since replicaPool is a List<ReplicaId>
+    // replicaByDcIterator will be a new iterator for the map
+    return otIterator;
```

Comment:
This `otIterator` is initialized in `SimpleOperationTracker` to a private class defined inside `SimpleOperationTracker`. We need to change this. The `otIterator` now should be initialized as an instance of `ParanoidDurabilityTrackerIterator`.